src/narjure/memory_management/persistence_manager.clj

```diff
@@ -2,18 +2,47 @@
   (:require
     [co.paralleluniverse.pulsar.actors :refer [self ! whereis cast! Server gen-server register! shutdown! unregister! set-state! state]]
     [narjure.actor.utils :refer [defactor]]
+    [narjure.memory-management.task-dispatcher :refer [c-map]]
+    [narjure.memory-management.concept-creator :refer [create-concept]]
+    [narjure.actor.utils :refer [read-seq-from-file]]
     [taoensso.timbre :refer [debug info]])
   (:refer-clojure :exclude [promise await]))
 
 (def aname :persistence-manager)
 
-(defn concept-state-request-handler
-  "Processes :concept-state-request-msg:
-    serialises concept state and returns
-    to requesting actor"
-  [from [msg]]
-  ()
-  (debug aname "process-concept-state-request-msg"))
+(defn persist-concept-state-handler
+  "Posts :concept-state-request-msg to each concept in c-map"
+  [from [_ path]]
+  (set-state! (assoc @state :path path :concept-count (count @c-map) :received-states 0))
+  (doseq [concept-ref (vals @c-map)]
+    (cast! concept-ref [:concept-state-request-msg]))
+  #_(debug aname "process-persist-concept-state-msg"))
+
+(defn concept-state-handler
+  "process each :concept-state-msg by serialising the state to backing store.
+   where state specifies the path of the backing store. The number of received
+   states is tracked. The file is overwritten."
+  [from [_ concept-state]]
+  (let [c-state (dissoc concept-state :active-concept-collator :general-inferencer :forgettable-concept-collator)]
+    (if (= (:received-states @state) 0)
```

Comment:
Oh, I see what you mean. This is what I meant by the code being messy here. If we pass in the state as parameters then we have to pass out the updated values and change the state in the calling function. Do you have a suggestion for how do so this efficiently?

 

 

From: Roman Volosovskyi [mailto:notifications@github.com] 
Sent: 30 April 2016 14:02
To: opennars/opennars2 <opennars2@noreply.github.com>
Cc: TonyLo1 <tony_lofthouse@btinternet.com>; Mention <mention@noreply.github.com>
Subject: Re: [opennars/opennars2] Actor update: concept-creator, forgettable-concept-collator, concept (#21)

 

In src/narjure/memory_management/persistence_manager.clj <https://github.com/opennars/opennars2/pull/21#discussion_r61669382> :

> -  (debug aname "process-concept-state-request-msg"))
> +(defn persist-concept-state-handler
> +  "Posts :concept-state-request-msg to each concept in c-map"
> +  [from [_ path]]
> +  (set-state! (assoc @state :path path :concept-count (count @c-map) :received-states 0))
> +  (doseq [concept-ref (vals @c-map)]
> +    (cast! concept-ref [:concept-state-request-msg]))
> +  #_(debug aname "process-persist-concept-state-msg"))
> +
> +(defn concept-state-handler
> +  "process each :concept-state-msg by serialising the state to backing store.
> +   where state specifies the path of the backing store. The number of received
> +   states is tracked. The file is overwritten."
> +  [from [_ concept-state]]
> +  (let [c-state (dissoc concept-state :active-concept-collator :general-inferencer :forgettable-concept-collator)]
> +    (if (= (:received-states @state) 0)

The global idea was to keep handlers as pure functions, so if this doesn't work here there is no value in adding state as parameter here.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub <https://github.com/opennars/opennars2/pull/21/files/8bd9a8b32e14e24ffc289103d9c215ab1c3875fc#r61669382>